### PR TITLE
feat: Various patches

### PR DESCRIPTION
**Pull request by Thomzoy**
## Handling intra-word linebreak as pollution
Adds a pollution pattern that detects intra-word linebreak, which can then be removed in the `get_text` method
## Qualifiers can process Span or Doc
This feature especially makes it easier to use such pipes in other pipes directly
## Add label_weights
Allows the user to mention label weights during span classification training
## Deduplicate spans between doc.ents and doc.spans during train
Previously, if the `span_getter` asks to get both `ents` and `spans`, entities would have been seen twice

## Checklist

- [x] If this PR is a bug fix, the bug is documented in the test suite.
- [x] Changes were documented in the changelog (pending section).
- [x] If necessary, changes were made to the documentation (eg new pipeline).

### CODE REVIEW

File: edsnlp/pipes/trainable/span_classifier/span_classifier.py [hunk 1]
```diff
@@ -165,6 +166,11 @@ class TrainableSpanClassifier(
         Name of the component
     embedding : SpanEmbeddingComponent
         The word embedding component
+    label_weights: Dict[str, Dict[Any, Number]]
```
Comment (percevalw) [inline edsnlp/pipes/trainable/span_classifier/span_classifier.py:169]: Can't use just use float ? `Number` also support complex, rationals numbers, etc

File: edsnlp/pipes/ner/behaviors/alcohol/patterns.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
     regex=[
         r"\balco[ol]",
         r"\bethyl",
-        r"(?<!(25.?)|(sevrage)).?\boh\b",
+        r"(?<!(25.{0,10}))\boh\b",
```
Comment (percevalw) [inline edsnlp/pipes/ner/behaviors/alcohol/patterns.py:6]: Doesn't this add false positives ?

File: edsnlp/pipes/trainable/span_classifier/span_classifier.py [hunk 3]
```diff
@@ -480,6 +516,9 @@ def forward(self, batch: SpanClassifierBatchInput) -> BatchOutput:
                         binding_scores[:, bindings_indexer],
                         batch["targets"][:, group_idx],
                         reduction="sum",
+                        weight=torch.tensor(self.label_weights)[bindings_indexer].to(
+                            binding_scores.device
```
Comment (percevalw) [inline edsnlp/pipes/trainable/span_classifier/span_classifier.py]: Do you have some results demonstrating an improvement in accuracy when weighting the loss ?

File: pyproject.toml [hunk 4]
```diff
@@ -260,6 +260,9 @@ where = ["."]
 "eds.span_linker"         = "edsnlp.pipes.trainable.span_linker.factory:create_component"
 "eds.biaffine_dep_parser" = "edsnlp.pipes.trainable.biaffine_dep_parser.factory:create_component"
 
+[project.entry-points."edsnlp_shedueles"]
```
Comment (percevalw) [inline pyproject.toml]:
```
[project.entry-points."edsnlp_schedules"]
```

File: edsnlp/utils/span_getters.py [hunk 5]
```diff
@@ -35,24 +35,24 @@
...
         yield doc[:]
         return
     if callable(span_getter):
         yield from span_getter(doc)
         return
+    seen = set()
     for key, span_filter in span_getter.items():
         if key == "*":
-            candidates = (span for group in doc.spans.values() for span in group)
+            candidates = [span for group in doc.spans.values() for span in group]
```
Comment (percevalw) [inline edsnlp/utils/span_getters.py:48]: This should stay a generator

File: edsnlp/utils/span_getters.py [hunk 6]
```diff
@@ -35,24 +35,24 @@
 ]
 
 
-def get_spans(doc, span_getter):
+def get_spans(doc, span_getter, deduplicate=True):
```
Comment (percevalw) [inline edsnlp/utils/span_getters.py:38]: Can we add a test for this new behavior ?

File: changelog.md [hunk 7]
```diff
@@ -3,6 +3,11 @@
 ## Unreleased
 
 ### Added
+- Handling intra-word linebreak as pollution : Adds a pollution pattern that detects intra-word linebreak, which can then be removed in the `get_text` method
+- Qualifiers can process `Span` or `Doc` : This feature especially makes it easier to use such pipes in other pipes directly
+- Various disorders/behaviors patches
+- Add label_weights : Allows the user to mention label weights during span classification training
+- Deduplicate spans between doc.ents and doc.spans during train : Previously, if the `span_getter` asks to get both `ents` and `spans`, entities would have been seen twice
```
Comment (percevalw) [inline changelog.md:10]:
```suggestion
- Handling intra-word linebreak as pollution : adds a pollution pattern that detects intra-word linebreak, which can then be removed in the `get_text` method
- Qualifiers can process `Span` or `Doc` : this feature especially makes it easier to nest qualifiers components in other components
- New label_weights parameter in eds.span_classifier`, which allows the user to set per label-value loss weights during training

### Fixed

- Various disorders/behaviors patches

### Changed

- Deduplicate spans between doc.ents and doc.spans during train: previously, a `span_getter` requesting entities from both `ents` and `spans` could yield duplicates.
```

File: edsnlp/utils/span_getters.py [hunk 8]
```diff
@@ -35,24 +35,24 @@
...
-        if span_filter is True:
-            yield from candidates
-        else:
-            for span in candidates:
-                if span.label_ in span_filter:
+        for span in candidates:
+            if (span_filter is True) or (span.label_ in span_filter):
+                if span not in seen:
                     yield span
+                    seen = seen | set([span]) if deduplicate else seen
```
Comment (percevalw) [inline edsnlp/utils/span_getters.py:55]:
```suggestion
if deduplicate:
    seen.add(span)
```

**Review by percevalw — COMMENTED** [review]
Thanks ! I've left a few comments

File: edsnlp/utils/doc_to_text.py [hunk 9]
```diff
@@ -99,6 +99,12 @@ def aggregate_tokens(
                     offset += len(part)
                     ends[i] = offset
             else:
+                if i > 0 and tag_hash == space_hash:
+                    if text_parts[i - 1][-1:] and (
+                        text_parts[i - 1][-1:] not in (" ", "\n")
+                    ):
+                        text_parts[i - 1] += " "
+                        offset += 1
```
[the remark this answers is not in the document]
Comment (Thomzoy) [reply edsnlp/utils/doc_to_text.py:107]: This (in conjonction with the additional [tokenizer exception](https://github.com/aphp/edsnlp/pull/391/files#diff-bc18124497162bedd7c1bbf5a4656b94470aa4500eadc10b03ec49ba920b9a8aL44) allows discarding intre-word linebreak of the form `diab-\ète`. For instance:

```python
import edsnlp, edsnlp.pipes as eds
from edsnlp.utils.doc_to_text import get_text

nlp = edsnlp.blank("eds")
nlp.add_pipe(
    eds.normalizer(
        accents=True,
        lowercase=True,
        quotes=True,
        spaces=True,
        pollution=True,
    )
)
text = "Le patient a un diab-\nète de type II."
doc = nlp(text)
```

Previous behavior:

```python
repr(get_text(doc, ignore_excluded=True))
# Out: 'Le patient a un diab-\nète de type II.'
```

New proposed behavior:

```python
repr(get_text(doc, ignore_excluded=True))
# Out: 'Le patient a un diabète de type II.'
```

File: edsnlp/pipes/qualifiers/family/family.py [hunk 10]
```diff
@@ -187,7 +187,8 @@ def set_extensions(self) -> None:
         if not Doc.has_extension("family"):
             Doc.set_extension("family", default=[])
 
-    def process(self, doc: Doc) -> FamilyResults:
+    def process(self, doc_like: Union[Doc, Span]) -> FamilyResults:
+        doc = self.ensure_doc(doc_like)
         matches = self.get_matches(doc)
```
Comment (percevalw) [inline edsnlp/pipes/qualifiers/family/family.py:192]: @Thomzoy Isn't this misleading, wrt the Doc|Span parameter type ? This function claims it can process a Span, so users might assume it only processes entities inside the span, when in reality it performs the matching on the full document, regardless of its doc_like parameter

File: edsnlp/pipes/ner/disorders/base.py [hunk 11]
```diff
@@ -113,8 +113,17 @@ def __call__(self, doc: Doc) -> Doc:
             annotated spaCy Doc object
         """
         spans = list(self.process(doc))
+        all_detailed_status = set(self.detailed_status_mapping.keys())
         for span in spans:
-            span._.detailed_status = self.detailed_status_mapping[span._.status]
+            if span._.status is not None and span._.status not in all_detailed_status:
```
[the remark this answers is not in the document]
Comment (percevalw) [reply edsnlp/pipes/ner/disorders/base.py:118]: Also, why don't we just:
1. check if `span._.status is None`, in which case `span._.detailed_status` will be None
2. just do `span._.detailed_status = self.detailed_status_mapping[span._.status]` and expect this to raise a KeyError if the status is incorrect
?

In addition, we could put all of this in a getter in `Span.set_extension("detailed_status", ...)`

File: edsnlp/pipes/ner/behaviors/alcohol/patterns.py (hunk 2 above)
Comment (Thomzoy) [reply edsnlp/pipes/ner/behaviors/alcohol/patterns.py:6]: Apparently no from the research project that suggested this change

File: edsnlp/utils/span_getters.py (hunk 6 above)
Comment (Thomzoy) [reply edsnlp/utils/span_getters.py:38]: Yes good call !

File: edsnlp/pipes/ner/disorders/base.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Thomzoy) [reply edsnlp/pipes/ner/disorders/base.py:118]: Not sure if we can do `Span.set_extension` since the mapping is pipe-dependent ?

File: edsnlp/pipes/qualifiers/family/family.py (hunk 10 above)
Comment (percevalw) [reply edsnlp/pipes/qualifiers/family/family.py:192]: Ah yes, I mixed `span.as_doc()` vs `span.doc`.
I think this poses another problem : `span.as_doc()` creates a new document, detached from the original doc, so any new attribute added won't be saved. If this feature actually used ? If not, I'd rather remove it at the moment, as this changes EDS-NLP API and makes it non uniform across our pipes (somes pipes accept both Span and Doc, some only Doc)

Comment (Thomzoy) [reply edsnlp/pipes/qualifiers/family/family.py:192]: Yes it was the exact goal : *Normal* use of qualifiers will still be through the `__call__` method, but if you want to qualify part of a document (for instance, an `assign` span from the ContextualMatcher) without *polluting* the initial document, you'll use `process`.

And in the end the `__call__` method stays consistent (only accepting Doc)

File: edsnlp/language.py [hunk 12]
```diff
@@ -41,6 +41,15 @@ class EDSLanguage(French):
...
+TOKENIZER_EXCEPTIONS = [
+    r"Dr\.",
+    r"Pr\.",
+    r"M\.",
+    r"Mme\.",
+    r"Mlle\.",
+    r"(?i:(?:ep\.))",
+    r"(?<![\W\d_])-\n",
+]
 TOKENIZER_EXCEPTIONS = [r"Dr\.", r"Pr\.", r"M\.", r"Mme\.", r"Mlle\.", r"(?i:(?:ep\.))"]
```
Comment (percevalw) [inline edsnlp/language.py:53]: @Thomzoy
```suggestion
TOKENIZER_EXCEPTIONS = [r"Dr\.", r"Pr\.", r"M\.", r"Mme\.", r"Mlle\.", r"(?i:(?:ep\.))"]
```
